### ISSUE TITLE
Probability 99% when pointing at a non-object

### ISSUE THREAD

**Issue by jordy-wijman** [apple/turicreate#1225]
So I've been trying out TuriCreate for some time now and it's actually working really good. There's only one issue I'm having with the image classification of TuriCreate. When I'm pointing at an actual object that I've trained, it works PERFECT, like it pretty much always gives the correct label back which is awesome. But when I point my camera to a non object (let's say a wall or the floor) it still recognizes it as an object and gives back a probability of 99%. We are making an app that recognizes objects in realtime and then places an AR label on the object in iOS.

We've trained 5 objects with aprox 300 images each and are using squeezenet as the pre-trained model. Please note that I've also tried using the standard pre-trained model, it pretty much gives the exact same issue's as written above.

Do you guys have any idea's on how we could get this percentage a little bit lower on non-objects? If there's any missing information, please let me know and I will edit this post to add more information!

`tc.image_classifier.create(train_data, target='label', max_iterations=1500, model='squeezenet_v1.1')`

**Comment by TobyRoseman** [comment]
Hi @jordy-wijman -

I think this is a very common issue with a lot of machine learning algorithms. Machine learning models are often over confident when give randomly generated examples or even just examples from categories outside of their training set. 

One thing you could try is to create another class in your training data called "other". Put lots of randomly picked images in that class. Or if walls and floors are common false positives for your case, put images of walls and floors in your class. Then hopefully when your model sees something which is not one of the five objects you care about, it will classify it as "other".

Hope this helps. Let us know if you have any additional questions. If you try my suggestion, let us know how it works.

**Comment by jordy-wijman** [comment]
Hi @TobyRoseman thanks a lot for your response, I'll definitely try that out tomorrow. Do you think that this would impact the probabilities a lot on the objects it knows? 

And I was also wondering if it would matter if I would train one object with let's say 800 images, and another one with 300. Would the model be more likely to favour the one with the 800 images over the one with the 300 in the list of probabilities? (Basically, would I still be able to get 99% probability if it's actually pointing at the object with the 300 image item?)

And one last question, I've also trained some models using TensorFlow, I did use other pre-trained models. But the probabilities I get from these models are lower in general with the exact same dataset. TuriCreate sometimes even hits 100%, where as with TensorFlow 90% would be considered very high. I also seemed to get less high scores on objects such as roofs and floors etc. The problem I have with TensorFlow though is that the results just aren't as good as TuriCreate when pointing at the actual object. TuriCreate always seems to recognize it, where TensorFlow (using MobileNet as pre-trained model) just isn't as accurate. Is there a reason why TensorFlow and TuriCreate are giving so much different results with the exact same dataset? Is it just because of the pre-trained model I picked?

**Comment by jordy-wijman** [comment]
Alright, so I've tested it I walked around the office for 2 minutes and made a video and used that content. It's actually working really well, so thanks a lot for the idea. The confidence on the objects are still as high as they were before, so we're happy with the results.

Here is a video that previews that it actually works really good https://streamable.com/h0w7s .
About the feature request, what would the solution in TuriCreate be? would it just take the lowest amount of pictures from a certain object, and limit the others to that amount?

**Comment by TobyRoseman** [comment]
Neat demo video. Thanks for sharing and thanks for letting us know how well the "other" category approach worked.

The feature request would be to add a `class_weights` parameter to `turicreate.image_classifier.create`. The image classifier uses logistic regression (see my previous link for details). So the image classifier would basically just pass that parameter to logistic regression. See the [API docs for logistic regression](https://apple.github.io/turicreate/docs/api/generated/turicreate.image_classifier.create.html#turicreate.image_classifier.create) for details about that parameter.

I'm going to close this issue, but let me know if you have any more questions.
